core/primitives/src/sharding.rs

```diff
@@ -98,20 +99,239 @@ pub struct ShardChunkHeader {
     pub hash: ChunkHash,
 }
 
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Clone, PartialEq, Eq, Debug)]
+#[borsh_init(init)]
+pub struct ShardChunkHeaderV2 {
```

Comment:
given that the struct is exactly the same as `ShardChunkHeaderV1`, do we really need to have a separate struct? Is it possible to infer how to check/compute hash based on the protocol version?

Reply:
This is a good point, though I think differentiating explicitly by a new struct is good because it means the `init` method used by Borsh can differ between the two. The issue is that the hash is set immediately upon deserialization, and the protocol version is not known at that point.